scarb/tests/tree.rs

```diff
@@ -0,0 +1,441 @@
+use indoc::indoc;
+
+use scarb_test_support::command::Scarb;
+use scarb_test_support::project_builder::{Dep, DepBuilder, ProjectBuilder};
+use scarb_test_support::workspace_builder::WorkspaceBuilder;
+
+#[test]
+fn simple() {
+    let t = TempDir::new().unwrap();
+
+    let dep1 = t.child("dep1");
+    ProjectBuilder::start()
+        .name("dep1")
+        .version("0.1.0")
+        .build(&dep1);
+
+    let dep2 = t.child("dep2");
+    ProjectBuilder::start()
+        .name("dep2")
+        .version("0.1.0")
+        .build(&dep2);
+
+    let root = t.child("root");
+    ProjectBuilder::start()
+        .name("root")
+        .version("0.1.0")
+        .dep("dep1", &dep1)
+        .build(&root);
+
+    Scarb::quick_snapbox()
+        .arg("tree")
+        .current_dir(&root)
+        .assert()
+        .success()
+        .stdout_matches(indoc! {r#"
+            root v0.1.0 ([..])
+            └── dep1 v0.1.0 ([..])
+        "#});
+}
+
+#[test]
+fn json_output() {
+    let t = TempDir::new().unwrap();
+
+    let dep1 = t.child("dep1");
+    ProjectBuilder::start()
+        .name("dep1")
+        .version("0.1.0")
+        .build(&dep1);
+
+    let dep2 = t.child("dep2");
+    ProjectBuilder::start()
+        .name("dep2")
+        .version("0.1.0")
+        .build(&dep2);
+
+    let root = t.child("root");
+    ProjectBuilder::start()
+        .name("root")
+        .version("0.1.0")
+        .dep("dep1", &dep1)
+        .build(&root);
+
+    Scarb::quick_snapbox()
+        .arg("--json")
+        .arg("tree")
+        .current_dir(&root)
+        .assert()
+        .success()
+        .stdout_matches(indoc! {r#"
+        [{"package":"root 0.1.0 ([..])","branches":[{"package":"dep1 0.1.0 ([..])"}]}]
+        "#});
+}
+
+#[test]
+fn requires_workspace() {
+    let t = TempDir::new().unwrap();
+    Scarb::quick_snapbox()
+        .env("RUST_BACKTRACE", "0")
+        .arg("tree")
+        .current_dir(&t)
+        .assert()
+        .failure()
+        .stdout_matches(indoc! {r#"
+            error: failed to read manifest at: [..]/Scarb.toml
+
+            Caused by:
+                [..]
+        "#});
+}
+
+#[test]
+fn no_dedupe() {
+    let t = TempDir::new().unwrap();
+
+    let common = t.child("common");
+    ProjectBuilder::start()
+        .name("common")
+        .version("0.1.0")
+        .build(&common);
+
+    let dep1 = t.child("dep1");
+    ProjectBuilder::start()
+        .name("dep1")
+        .version("0.1.0")
+        .dep("common", &common)
+        .build(&dep1);
+
+    let dep2 = t.child("dep2");
+    ProjectBuilder::start()
+        .name("dep2")
+        .version("0.1.0")
+        .dep("common", &common)
+        .build(&dep2);
+
+    let root = t.child("root");
+    ProjectBuilder::start()
+        .name("root")
+        .version("0.1.0")
+        .dep("dep1", &dep1)
+        .dep("dep2", &dep2)
+        .build(&root);
+
+    Scarb::quick_snapbox()
+        .arg("tree")
+        .arg("--no-dedupe")
```

Comment:
removing the arg does not change the output 🤔 Please add some deps to common, so they can be deduplicated.

Reply:
2d888d7353f2c0adbb9538bfdb68a7a12aa6799a